Notes/Proposals/Explicit-partial-type-inference.md

```diff
@@ -0,0 +1,313 @@
+
+```csharp
+Wrapper<_> wrapper = ... // I get an wrapper, which I'm interested in, but I don't care about the type arguments, because I don't need them in my code.
+wrapper.DoSomething();
+```
+At the end, casting could use the `_` placeholder as well.
+
+```csharp
+Foo<int, string> myvar = (Foo<_,_>)myobject; // Hint the type arguments based on target or other potential source of type information like default or named type arguments.
+```
+
+An interesting thing would be to allow the `_` placeholder in member lookup as you can see in the example below.
+On the first see, it can look wierd.
+
+```csharp
+static class C1<T1> {
+    static class C2 <T2> {
+        public (T1, T2) Foo(T1 t, T2 t) {}
+    }
+}
+
+var a = C1<_>.C2<_>.Foo(1, 1);
+```
+
+But in combination with default parameters, It might be useful in cases, where we use entity as a global provider of something, which we determine by type.
+
+```csharp
+static class Factory<T = Default> {
+    public static T Create(){...}
+}
+
+int a = Factory<_>.Create(); // Calls Factory<int>.Create();
+var b = Factory<_>.Create(); // Calls Factory<Default>.Create();
+```
+
+Although it is unlikely that it would be added into C#, we would like to investigate what should be done to enable it for a deeper understanding of type inference in Roslyn.
+
+Based on mentioned advantages that `_` placeholder can bring to type inference we feel to add it to C# worthly.
+
+## Scope
+
+Partial type inference can be solved in various ways.
+We chose a feature enabling to hint the compiler by specifying ambitious type arguments and letting the remaining ones on him.
+It aims at cases, where we want just to specify the arity of desired generic method(type) or specify a parameter that is not possible to infer from the context but let the compiler infer the remaining arguments. 
+
+## Design
+
+As we mentioned in the motivation, we see the usage of our proposal in the following places.
+
+1. method call
+2. object construction
+3. variable declaration
+4. casting
+
+### Method call
+
+We would like to somehow mark inferred type arguments. 
+The most natural approach would be to replace the type arguments with some keyword. 
+Let's pick `_` for it for now and show it in the examples.
+
+```csharp
+void foo<T1, T2, T3>(T2 arg) {}
+
+foo<int, _, int>("string"); // _ is marked as "inferred by compiler"
+```
+
+One can suggest not marking it and just leaving it on the compiler. 
+However, it would bring ambiguities between which arguments should be inferred. 
+
+```csharp
+void foo(int arg1, int arg2) {}
+void foo<T1>(T1 arg1, int arg2) {}
+void foo<T1, T2>(T1 arg1, T2 arg2) {}
+
+foo(1,2); // We don't know which overload to choose
+foo<>(1,2); // Still, it is not clear which generic overload should be choose.
+```
+
+Because of mentioned issues, we believe that marking each type argument, what the compiler will infer, is the right way how to do it.
+To make it possible, we would have to change C# grammar
```

Comment:
It will be useful to link to the section of the C# standard this grammar was obtained from, and to use ```` ```diff ```` notation to indicate which parts of the grammar are being modified by this proposal.